File: WebWindowNetCore/WebViewForm.fs
namespace WebWindowNetCore
#if Windows
open System
open System.Drawing
open System.Runtime.InteropServices
open System.Windows.Forms
open Microsoft.Web.WebView2.Core
open Microsoft.Web.WebView2.WinForms
open FSharpTools
open System.ComponentModel

type WebViewForm(appDataPath: string, settings: WebViewBase) as this = 
    inherit Form()

    let webView = new WebView2()

    do 
        (webView :> ComponentModel.ISupportInitialize).BeginInit()
        this.SuspendLayout();
        webView.AllowExternalDrop <- true
        webView.CreationProperties <- null
        webView.DefaultBackgroundColor <- Color.White
        webView.Location <- Point (0, 0)
        webView.Margin <- Padding 0
        webView.Dock <- DockStyle.Fill
        webView.TabIndex <- 0
        webView.ZoomFactor <- 1

        //OnFormCreation.Invoke(this);

        settings.ResourceIconValue 
        |> Option.iter (fun i -> 
                                Resources.get i
                                |>Option.iter (fun s -> this.Icon <- new Icon (s)))

        this.AutoScaleDimensions <- SizeF(8F, 20F)
        this.AutoScaleMode <- AutoScaleMode.Font

        let bounds = Bounds.retrieve settings.AppIdValue
        if bounds.X.IsSome && bounds.Y.IsSome then
            this.Location <- Point(bounds.X |> Option.defaultValue 0, bounds.Y |> Option.defaultValue 0)
        this.Size <- Size(bounds.Width |> Option.defaultValue settings.WidthValue, bounds.Height |> Option.defaultValue settings.HeightValue)
        this.WindowState <- if bounds.IsMaximized then FormWindowState.Maximized else FormWindowState.Normal

        if settings.SaveBoundsValue then
            this.FormClosing.Add(this.onClosing)

        this.Load.Add(this.onLoad)

        this.Text <- settings.TitleValue
        this.Controls.Add webView

        (webView :> ComponentModel.ISupportInitialize).EndInit ()
        this.ResumeLayout false

        if settings.Requests |> List.length > 0 then
            Server.start settings

        async {
            let! enf = CoreWebView2Environment.CreateAsync(
                        null, 
                        appDataPath, 
                        CoreWebView2EnvironmentOptions(
                            customSchemeRegistrations = ResizeArray<CoreWebView2CustomSchemeRegistration> [ 
                                CoreWebView2CustomSchemeRegistration("res") 
                            ] ))
                        |> Async.AwaitTask
            do! webView.EnsureCoreWebView2Async(enf) |> Async.AwaitTask
            webView.CoreWebView2.AddWebResourceRequestedFilter("res:*", CoreWebView2WebResourceContext.All)
            webView.CoreWebView2.AddHostObjectToScript("Callback", Callback(this))
            webView.CoreWebView2.WebResourceRequested.Add(this.serveRes)
            webView.CoreWebView2.Settings.AreBrowserAcceleratorKeysEnabled <- false
            webView.CoreWebView2.Settings.IsPasswordAutosaveEnabled <- true
            webView.CoreWebView2.Settings.AreDefaultContextMenusEnabled <- settings.DefaultContextMenuDisabledValue = false
            webView.CoreWebView2.WindowCloseRequested.Add(fun _ -> this.Close())
            webView.CoreWebView2.ContainsFullScreenElementChanged.Add(this.onFullscreen)

            webView.Source <- Uri (settings.GetUrl ())
            webView.ExecuteScriptAsync(@"
                const callback = chrome.webview.hostObjects.Callback
            ") 
                |> Async.AwaitTask
                |> ignore

            webView.ExecuteScriptAsync(Requests.getScript settings.RequestPortValue true) 
                |> Async.AwaitTask
                |> ignore
        } 
        |> Async.StartWithCurrentContext 

    member this.WebView = webView

    member this.MaximizeWindow () = this.WindowState <- FormWindowState.Maximized
    member this.MinimizeWindow() = this.WindowState <- FormWindowState.Minimized
    member this.RestoreWindow() = this.WindowState <- FormWindowState.Normal
    member this.ShowDevtools () = 
        if settings.DevToolsValue then
            webView.CoreWebView2.OpenDevToolsWindow()
    member this.GetWindowState() = (int)this.WindowState

    member this.onLoad (_: EventArgs) =
        let bounds = Bounds.retrieve settings.AppIdValue
        if bounds.X.IsSome && bounds.Y.IsSome 
                && Screen.AllScreens |> Seq.exists (fun s -> s.WorkingArea.IntersectsWith(Rectangle(
                                                                                                            bounds.X |> Option.defaultValue 0, 
                                                                                                            bounds.Y |> Option.defaultValue 0, 
                                                                                                            this.Size.Width, 
                                                                                                            this.Size.Height))) then
            this.Location <- Point(bounds.X |> Option.defaultValue 0 , bounds.Y |> Option.defaultValue 0)
            this.WindowState <- if bounds.IsMaximized then FormWindowState.Maximized else FormWindowState.Normal
        ()
    member this.onClosing (e: FormClosingEventArgs) =
            { Bounds.retrieve settings.AppIdValue with 
                X = if this.WindowState = FormWindowState.Maximized then Some this.RestoreBounds.Location.X else Some this.Location.X
                Y = if this.WindowState = FormWindowState.Maximized then Some this.RestoreBounds.Location.Y else Some this.Location.Y
                Width = if this.WindowState = FormWindowState.Maximized then Some this.RestoreBounds.Size.Width else Some this.Size.Width
                Height = if this.WindowState = FormWindowState.Maximized then Some this.RestoreBounds.Size.Height else Some this.Size.Height
                IsMaximized = this.WindowState = FormWindowState.Maximized }
            |> Bounds.save settings.AppIdValue

    member this.onFullscreen _ =
        if webView.CoreWebView2.ContainsFullScreenElement then
            this.TopMost <- true
            this.FormBorderStyle <- FormBorderStyle.None
            this.WindowState <- FormWindowState.Maximized
            Taskbar.hide ()
        else
            this.TopMost <- false
            this.WindowState <- FormWindowState.Normal
            this.FormBorderStyle <- FormBorderStyle.Sizable
            Taskbar.show ()

    member this.serveRes e = 
        let serveResourceStream (url: string) (stream: System.IO.Stream) = 
            try
                e.Response <- this.WebView.CoreWebView2.Environment.CreateWebResourceResponse(stream, 200, "OK", sprintf "Content-Type: %s" (ContentType.get url))
            with
            | _ ->  e.Response <- this.WebView.CoreWebView2.Environment.CreateWebResourceResponse(null, 404, "Not Found", "")

        let uri = 
            Uri.UnescapeDataString(e.Request.Uri)
            |> String.substring 6

        uri
        |> Resources.get 
        |> Option.iter (serveResourceStream uri)
        ()

    override this.OnClosing(e: CancelEventArgs) = 
        base.OnClosing(e)
        settings.CanCloseValue
        |> Option.iter (fun cc -> e.Cancel <- cc() = false)
        
and [<ComVisible(true)>] Callback(parent: WebViewForm) =

    member this.ShowDevtools() = parent.ShowDevtools()
    member this.MaximizeWindow() = parent.MaximizeWindow()
    member this.MinimizeWindow() = parent.MinimizeWindow()
    member this.RestoreWindow() = parent.RestoreWindow()
    member this.GetWindowState() = parent.GetWindowState()
    
    // public Task DragStart(string fileList)
    //     => JsonSerializer.Deserialize<FileListType>(fileList, JsonWebDefaults)
    //         .Map(flt =>
    //             parent.DragStart(flt!.Path, flt!.FileList));

    // member this.ScriptAction(id: int, msg: string) = parent.ScriptAction(id, msg)
#endif


    

File: WebWindowNetCore/Windows/WebViewForm.cs
#if Windows
using System.Runtime.InteropServices;
using System.Text;
using ClrWinApi;
using CsTools;
using CsTools.Extensions;
using Microsoft.Web.WebView2.Core;
using Microsoft.Web.WebView2.WinForms;

namespace WebWindowNetCore.Windows;

class WebViewForm : Form
{
    public WebView2 WebView { get; } = new();

    public WebViewForm(string appDataPath, WebView settings)
    {
        saveBounds = settings.saveBounds;
        appId = settings.appId;
        canClose = settings.canClose;
        devTools = settings.devTools;
        width = settings.width;
        height = settings.height;
        withoutNativeTitlebar = settings.withoutNativeTitlebar;
        settings.Initialize(this);
        SuspendLayout();
        WebView.AllowExternalDrop = true;
        WebView.CreationProperties = null;
        if (settings.backgroundColor.HasValue)
            WebView.DefaultBackgroundColor = settings.backgroundColor.Value;
        WebView.Dock = DockStyle.Fill;
        WebView.TabIndex = 0;
        WebView.ZoomFactor = 1;
        settings.onformCreate?.Invoke(this);

        if (settings.resourceIcon != null)
            Icon = new Icon(Resources.Get(settings.resourceIcon)!);
        AutoScaleMode = AutoScaleMode.Font;

        if (settings.saveBounds)
        {
            var bounds = settings.saveBounds
                ? WebWindowNetCore.Bounds.Retrieve(settings.appId)
                : null;
            Size = new Size(bounds?.Width ?? settings.width, bounds?.Height ?? settings.height);
            WindowState = bounds?.IsMaximized == true ? FormWindowState.Maximized : FormWindowState.Normal;
        }
        isMaximized = WindowState == FormWindowState.Maximized;

        if (settings.saveBounds)
            FormClosing += OnClose;
        if (settings.canClose != null)
            FormClosing += OnCanClose;
        HandleCreated += OnHandle;
        Load += OnLoad;
        Text = settings.title;

        panel.Dock = DockStyle.Fill;
        panel.Controls.Add(WebView);
        Controls.Add(panel);

        ResumeLayout(false);

        Init();
        async void Init()
        {
            var env = await CoreWebView2Environment.CreateAsync(null, appDataPath, new CoreWebView2EnvironmentOptions(
                            customSchemeRegistrations: [
                                new CoreWebView2CustomSchemeRegistration("res")
                                {
                                    //TreatAsSecure = true,
                                    HasAuthorityComponent = true
                                }
                            ], additionalBrowserArguments: settings.withoutNativeTitlebar ? "--enable-features=msWebView2EnableDraggableRegions" : ""));
            await WebView.EnsureCoreWebView2Async(env);
            WebView.CoreWebView2.AddWebResourceRequestedFilter("res:*", CoreWebView2WebResourceContext.All);
            WebView.CoreWebView2.WebResourceRequested += ServeRes;
            WebView.CoreWebView2.Settings.AreBrowserAcceleratorKeysEnabled = false;
            WebView.CoreWebView2.Settings.IsPasswordAutosaveEnabled = true;
            WebView.CoreWebView2.Settings.AreDefaultContextMenusEnabled = settings.defaultContextMenuDisabled == false;
            WebView.CoreWebView2.Settings.IsWebMessageEnabled = true;
            WebView.CoreWebView2.WebMessageReceived += WebMessageReceived;
            WebView.CoreWebView2.ContainsFullScreenElementChanged += OnFullscreen;
            WebView.CoreWebView2.WindowCloseRequested += (s, e) => Close();

            WebView.Source = new Uri(settings.GetUrl());

            await WebView.ExecuteScriptAsync(WebWindowNetCore.Windows.ScriptInjection.Get(settings.title));
            await Task.Delay(50);
            settings.RunJavascript($"WEBVIEWsetMaximized({(isMaximized ? "true" : "false")})");
            if (settings.withoutNativeTitlebar)
                Resize += (s, e) =>
                    {
                        if (WindowState == FormWindowState.Maximized != isMaximized)
                            isMaximized = WindowState == FormWindowState.Maximized;
                        settings.RunJavascript($"WEBVIEWsetMaximized({(isMaximized ? "true" : "false")})");
                    };
        }
    }

    void OnClose(object? _, FormClosingEventArgs e)
    {
        var bounds = (saveBounds
            ? WebWindowNetCore.Bounds.Retrieve(appId)
            : new Bounds(null, null, null, null, false))
            with {
            X = WindowState == FormWindowState.Maximized ? RestoreBounds.Location.X : Location.X,
            Y = WindowState == FormWindowState.Maximized ? RestoreBounds.Location.Y : Location.Y,
            Width = WindowState == FormWindowState.Maximized ? RestoreBounds.Size.Width : Size.Width,
            Height = WindowState == FormWindowState.Maximized ? RestoreBounds.Size.Height : Size.Height,
            IsMaximized = WindowState == FormWindowState.Maximized
        };
        WebWindowNetCore.Bounds.Save(appId, bounds);
    }

    void OnLoad(object? _, EventArgs e)
    {
        var bounds = WebWindowNetCore.Bounds.Retrieve(appId);
        if (bounds.X.HasValue && bounds.Y.HasValue)
            Location = new Point(bounds.X ?? 0, bounds.Y ?? 0);

        var screenBounds = Screen.FromRectangle(Bounds).WorkingArea;
        if (!screenBounds.Contains(Bounds))
        {
            // Adjust the form's location if it is out of bounds
            Location = new Point(Math.Max(screenBounds.X, Bounds.X), Math.Max(screenBounds.Y, Bounds.Y));
            // Ensure the form fits within the screen
            Size = new Size(Math.Min(screenBounds.Width, Bounds.Width), Math.Min(screenBounds.Height, Bounds.Height));
        }
        if (bounds.X.HasValue
                && bounds.Y.HasValue
                && Screen.AllScreens.Any(s => s.WorkingArea.IntersectsWith(new Rectangle(bounds.X ?? 0, bounds.Y ?? 0, Size.Width, Size.Height))))
            Location = new Point(bounds.X ?? 0, bounds.Y ?? 0);
        Size = new Size(bounds.Width ?? width, bounds.Height ?? height);
    }

    void OnCanClose(object? _, FormClosingEventArgs e)
    {
        if (canClose?.Invoke() == false)
            e.Cancel = true;
    }

    void OnHandle(object? _, EventArgs e)
    {
        Theme.StartDetection(SetDarkMode);
        SetDarkMode(Theme.IsDark());
    }

    Unit SendTextResponse(int code, string status, string text, CoreWebView2WebResourceResponse response)
        => Unit
            .Value
            .SideEffect(_ =>
            {
                var stream = new MemoryStream(Encoding.UTF8.GetBytes(text));
                response = WebView.CoreWebView2.Environment.CreateWebResourceResponse(stream, code,
                        status, $"Content-Type: text/plain");
            });

    Unit SendNotFound(CoreWebView2WebResourceResponse response)
        => SendTextResponse(404, "Not Found", "Resource not found", response);

    void ServeRes(object? _, CoreWebView2WebResourceRequestedEventArgs e)
    {
        try
        {
            var uri = "/" + Uri.UnescapeDataString(e.Request.Uri)[6..].SubstringAfter('/').SubstringUntil('?');
            uri = uri != "/" ? uri : "/index.html";
            var stream = Resources.Get(uri);

            if (stream != null)
            {
                try
                {
                    e.Response = WebView.CoreWebView2.Environment.CreateWebResourceResponse(stream, 200,
                        "OK", $"Content-Type: {uri.GetFileExtension()?.ToMimeType() ?? "text/html"}\nAccess-Control-Allow-Origin: *");
                }
                catch
                {
                    SendNotFound(e.Response);
                }
            }
        }
        catch
        {
            SendNotFound(e.Response);
        }
    }

    void WebMessageReceived(object? _, CoreWebView2WebMessageReceivedEventArgs e)
    {
        var msg = e.TryGetWebMessageAsString();
        if (msg == "maximize")
            WindowState = FormWindowState.Maximized;
        else if (msg == "minimize")
            WindowState = FormWindowState.Minimized;
        else if (msg == "restore")
            WindowState = FormWindowState.Normal;
        else if (msg == "close")
            Close();
    }

    void OnFullscreen(object? s, object _)
    {
        if (WebView.CoreWebView2.ContainsFullScreenElement)
        {
            TopMost = true;
            FormBorderStyle = FormBorderStyle.None;
            WindowState = FormWindowState.Maximized;
            //Taskbar.hide()
        }
        else
        {
            TopMost = false;
            WindowState = FormWindowState.Normal;
            FormBorderStyle = FormBorderStyle.Sizable;
            //Taskbar.show ()
        }
    }

    void SetDarkMode(bool dark)
        => Invoke(() =>
            {
                var _ = Api.DwmSetWindowAttribute(Handle, DwmWindowAttribute.ImmersiveDarkMode, [dark ? 1 : 0], 4);
                BackColor = dark ? Color.Black : Color.White;
            });

    protected override void WndProc(ref Message m)
    {
        if (DesignMode || !withoutNativeTitlebar)
            base.WndProc(ref m);
        else
            switch (m.Msg)
            {
                case WM_NCCALCSIZE:
                    CalcSizeNoTitlebar(ref m);
                    break;
                default:
                    base.WndProc(ref m);
                    break;
            }

        void CalcSizeNoTitlebar(ref Message m)
        {
            var isZoomedTop = Api.IsZoomed(Handle) ? 7 : 0;
            var isZoomedAll = Api.IsZoomed(Handle) ? 3 : 0;
            if (m.WParam != 0)
            {
                var nccsp = NcCalcSizeParams.Fromnint(m.LParam);
                nccsp.Rgrc0.Top += 1 + isZoomedTop;
                nccsp.Rgrc0.Bottom -= 5 + isZoomedAll;
                nccsp.Rgrc0.Left += 5 + isZoomedAll;
                nccsp.Rgrc0.Right -= 5 + isZoomedAll;
                Marshal.StructureToPtr(nccsp, m.LParam, true);
            }
            else
            {
                var clnRect = Marshal.PtrToStructure<Rect>(m.LParam);
                clnRect.Top += 1 + isZoomedTop;
                clnRect.Bottom -= 5 + isZoomedAll;
                clnRect.Left += 5 + isZoomedAll;
                clnRect.Right -= 5 + isZoomedAll;
                Marshal.StructureToPtr(clnRect, m.LParam, true);
            }
            m.Result = 0;
        }
    }

    bool isMaximized;
    const int WM_NCCALCSIZE = 0x83;
    readonly int width;
    readonly int height;
    readonly bool devTools;
    readonly Panel panel = new();
    readonly bool saveBounds;
    readonly string appId;
    readonly bool withoutNativeTitlebar;
    readonly Func<bool>? canClose;
}

#endif


    

